### ISSUE TITLE
Compiler crash using `main` in a package.d file

### ISSUE THREAD

**Issue by JinShil** [ldc-developers/ldc#2777]
board/package.d
```D
module board;

extern void main();
extern(C) void reset()
{
    main();
}
```

Compile
```
$ ldc2 -c -defaultlib= -conf= -Iruntime board/package.d
LLVMSymbolizer: error reading file: No such file or directory
#0 0x000000000289061a (ldc2+0x289061a)
#1 0x000000000288e8de (ldc2+0x288e8de)
#2 0x000000000288ea02 (ldc2+0x288ea02)
#3 0x00007f1e3b2c7a80 __restore_rt (/usr/lib/libpthread.so.0+0x11a80)
#4 0x0000000000c8729a (ldc2+0xc8729a)
#5 0x0000000000c90c3b (ldc2+0xc90c3b)
#6 0x0000000000c90eff (ldc2+0xc90eff)
#7 0x0000000000c8da7d (ldc2+0xc8da7d)
#8 0x0000000000c798d0 (ldc2+0xc798d0)
#9 0x0000000000c7b044 (ldc2+0xc7b044)
#10 0x0000000000c794d4 (ldc2+0xc794d4)
#11 0x0000000000c3eaab (ldc2+0xc3eaab)
#12 0x0000000000c33523 (ldc2+0xc33523)
#13 0x0000000000c36634 (ldc2+0xc36634)
#14 0x0000000000c55ec5 (ldc2+0xc55ec5)
#15 0x0000000000d01449 (ldc2+0xd01449)
#16 0x0000000000cd7226 (ldc2+0xcd7226)
#17 0x0000000000afa990 (ldc2+0xafa990)
#18 0x0000000000cd5007 (ldc2+0xcd5007)
#19 0x00000000029d5420 (ldc2+0x29d5420)
#20 0x00000000029d52a6 (ldc2+0x29d52a6)
#21 0x00007f1e3a97006b __libc_start_main (/usr/lib/libc.so.6+0x2306b)
#22 0x0000000000a22455 (ldc2+0xa22455)
Segmentation fault (core dumped)
```

Commenting out the call to `main` avoids the crash.
```D
module board;

extern void main();
extern(C) void reset()
{
    // main();
}
```

The above is just a reduced method to reproduce the error.  In my actual application I receive an additional error message "only one main function allowed"
```
$ ldc2 -c -Os -mtriple=thumb-none-eabi -float-abi=hard -mcpu=cortex-m4 -Isource/runtime source/board/random.d source/board/spi5.d source/board/lcd.d source/board/ltdc.d source/board/statusLED.d source/board/ILI9341.d source/board/package.d source/main.d source/stm32f42/pwr.d source/stm32f42/scb.d source/stm32f42/mmio.d source/stm32f42/dma2d.d source/stm32f42/spi.d source/stm32f42/bus.d source/stm32f42/flash.d source/stm32f42/ltdc.d source/stm32f42/rcc.d source/stm32f42/rng.d source/stm32f42/trace.d source/stm32f42/nvic.d source/stm32f42/gpio.d -of=binary/firmware.o
source/board/package.d(36): Error: only one main function allowed
LLVMSymbolizer: error reading file: No such file or directory
#0 0x000000000289061a (ldc2+0x289061a)
#1 0x000000000288e8de (ldc2+0x288e8de)
#2 0x000000000288ea02 (ldc2+0x288ea02)
#3 0x00007f7a125f8a80 __restore_rt (/usr/lib/libpthread.so.0+0x11a80)
#4 0x0000000000c8729a (ldc2+0xc8729a)
#5 0x0000000000c90c3b (ldc2+0xc90c3b)
#6 0x0000000000c90eff (ldc2+0xc90eff)
#7 0x0000000000c8da7d (ldc2+0xc8da7d)
#8 0x0000000000c798d0 (ldc2+0xc798d0)
#9 0x0000000000c7b044 (ldc2+0xc7b044)
#10 0x0000000000c7b044 (ldc2+0xc7b044)
#11 0x0000000000c794d4 (ldc2+0xc794d4)
#12 0x0000000000c3eaab (ldc2+0xc3eaab)
#13 0x0000000000c33523 (ldc2+0xc33523)
#14 0x0000000000c36634 (ldc2+0xc36634)
#15 0x0000000000c55ec5 (ldc2+0xc55ec5)
#16 0x0000000000d01449 (ldc2+0xd01449)
#17 0x0000000000cd7226 (ldc2+0xcd7226)
#18 0x0000000000afa990 (ldc2+0xafa990)
#19 0x0000000000cd5007 (ldc2+0xcd5007)
#20 0x00000000029d5420 (ldc2+0x29d5420)
#21 0x00000000029d52a6 (ldc2+0x29d52a6)
#22 0x00007f7a11ca106b __libc_start_main (/usr/lib/libc.so.6+0x2306b)
#23 0x0000000000a22455 (ldc2+0xa22455)
Segmentation fault (core dumped)
```

But I haven't found a way to reproduce that error in a reduced test yet.

**Comment by thewilsonator** [comment]
I seem to remember this happening before. It was to do with the fact that main is treated specially and is signature is altered resulting in a signature mismatch when called. Try to change the definition to the   usual `extern(C) int main(int, char**)` and call it with `main(0,null);`.

**Comment by thewilsonator** [comment]
As for the duplicate main definition, I'm not quite sure why it happens, perhaps DMDFE thinks that the forward declaration counts as one? Separate compilation should solve that issue. If not Try separate compilation with two definitions one of the the proper one, the other a dummy marked `@weak`.

**Comment by JinShil** [comment]
@thewilsonator Thanks!

As you alluded to, the following 2 variations seem to work.

**Test 2**
runtime/object.d
```D
module object;

alias string = immutable(char)[];
```

board/package.d
```D
module board;

int main(string[] args);
extern(C) void reset()
{
    main(null);
}
```

**Test 2**
board/package.d
```D
module board;

extern(C) int main(int, char**);
extern(C) void reset()
{
    main(0, null);
}
```

I still can't seem to get bast the "only one main function allowed" error, though, in my real application.  I'll keep investigating.

**Comment by JinShil** [comment]
> extern(C) void(string[] args) {}? that doesn't look valid.

Typo. Sorry.  I updated the comment.

**Comment by thewilsonator** [comment]
Try this `ldc/gen/functions.cpp` around line 600:
```diff
  // main
-  if (isMainFunction(fdecl)) {
+  if (isMainFunction(fdecl) && fdecl->fbody) { // ignore forward declarations
    // Detect multiple main functions, which is disallowed. DMD checks this
    // in the glue code, so we need to do it here as well.
    if (gIR->mainFunc) {
      error(fdecl->loc, "only one `main` function allowed");
    }
    gIR->mainFunc = func;
  }
```

**Comment by JinShil** [comment]
Finally figured out how to reproduce the "only one main function allowed" error message.  You need to add the `of=whatever` compiler switch.  If you omit it the error message doesn't occur.

runtime/object.d
```D
module object;

alias string = immutable(char)[];
```
board/package.d
```D
module board;

void main(string[] args);
void reset()
{
    main(null);
}
```

main.d
```D
module main;

void main(string[] args) { }
```

Compile:
```
$ ldc2 -c -Iruntime board/package.d main.d -of=a.out
board/package.d(3): Error: only one main function allowed
```

If you remove `-of=a.out` the error does not occur.

**Comment by thewilsonator** [comment]
Hmm, bizzarre.

**Comment by kinke** [comment]
Nicholas nailed it, it's an overzealous check for multiple C/D mains in a single object file, incl. declarations.

**Comment by kinke** [comment]
@JinShil: Does your code work now with proper signature and manual linking? Just trying to anticipate further issues in your interest, as the next release may be 1.11 final.

**Comment by JinShil** [comment]
> Does your code work now with proper signature and manual linking?

I don't quite understand your explanation above yet, but if I use `extern(C) void main(string[] args)` for my `main` implementation I can avoid the crash and the `only one main function allowed` error function allowed error message no longer occurs.

I'm still running into issues that appear unrelated to this issue.  Here's a small list

### Issue 1
If I use the following for my object.d file...
```D
module object;

alias size_t    = typeof(int.sizeof);
alias ptrdiff_t = typeof(cast(void*)0 - cast(void*)0);

alias string = immutable(char)[];

version(GNU)
{
    class Object
    { }
}
```
... I get several errors like
```
source/stm32f42/bus.d(20): Error: class `stm32f42.bus.CorePeripherals` unable to resolve forward reference in definition
```
The code where this occurs looks like
```D
module stm32f42.bus;

final abstract class CorePeripherals
{
    static @property uint address()
    {
        return 0xE000_0000;
    }
}
```
If I comment out the `class Object {}` implementation, the error disappears.  What bothers me about this is that it is enclosed in a `version(GNU)` block, so I don't think it should matter.  However, it may become a non-issue when I get a newer GDC compiler because then I should be able to remove `Object` from my code completely.  It's still a bug, though, IMO.

### Issue 2
GDC seems to use `@inline` attribute to force inlining (@ibuclaw, Will this still be the case with the next GDC release?).  LDC seems to use `pragma(inline, true)`.  I'm still trying to figure out how to write my code so I can force inlining with both compilers.

### Issue 3
I currently use `@naked` from GDC's attributes in one spot in my code.  I think I could use https://github.com/ldc-developers/ldc/pull/2773

### Issue 4
I have the following in my minimal runtime for `volatileLoad` and `volatileStore`:
```D
module core.bitop;

nothrow:
@safe:
@nogc:

ubyte  volatileLoad(ubyte * ptr);
ushort volatileLoad(ushort* ptr);
uint   volatileLoad(uint  * ptr);
ulong  volatileLoad(ulong * ptr);

void volatileStore(ubyte * ptr, ubyte  value);
void volatileStore(ushort* ptr, ushort value);
void volatileStore(uint  * ptr, uint   value);
void volatileStore(ulong * ptr, ulong value);
```
I get many linker errors like the following:
```
arm-none-eabi-ld: gpio.d:(.text.hardwareInit[hardwareInit]+0x46): undefined reference to `_D4core5bitop12volatileLoadFNbNiNfPkZk'
arm-none-eabi-ld: gpio.d:(.text.hardwareInit[hardwareInit]+0x58): undefined reference to `_D4core5bitop13volatileStoreFNbNiNfPkkZv'
```
I thought these were supposed to be compiler intrinsics so I didn't need to provide an implementation.  When I compile with GDC I do not get any errors.

I haven't filed any issues for these things yet because I'm still familiarizing myself with the LDC, and trying to upgrade my code for the latest D language changes, and it takes time for me to reduce these things to actionable issues that anyone can reproduce without my specific hardware.  

I'm also waiting on the latest GDC release because the language itself may have changed since the last release of GDC and I may need to make accommodations for both compilers, anyway.

Bottom line is I have more work to do, and it will take me some time to get through it, and communicate results to others.  But, if I can do anything to be less annoying, let me know.

**Comment by ibuclaw** [comment]
> GDC seems to use @inline attribute to force inlining (@ibuclaw, Will this still be the case with the next GDC release?).

Yes, `@inline` is not going anywhere. The implementation of `pragma(inline)` for now is still sitting there in an open PR.

**Comment by kinke** [comment]
Issue 1: Baffling, will need some investigation.
Issue 2: `pragma(inline, true)` is an official 'predefined' pragma in the language. LDC doesn't do cross-module inlining by default, so inlining isn't guaranteed. See https://github.com/ldc-developers/ldc/issues/2552.
Issue 3: Does the function in question have no params? Then current #2773 will do without further hacking.
Issue 4: AFAIK, there are no magic intrinsics in LDC (i.e., detected from name/module only). See https://github.com/ldc-developers/druntime/blob/ldc/src/core/bitop.d#L920-L936 for how LDC's druntime declares the ones you need.

**Comment by JinShil** [comment]
> Does the function in question have no params?

Yes, no parameters.  It's the entry point to my embedded system (the first procedure to run the the hardware reset).

> how LDC's druntime declares the ones you need.

I think those should be `nothrow`, and could potentially also be `@safe` and `@nogc`.  I think that would resolve the linker errors.

**Comment by kinke** [comment]
> I think those should be nothrow, and could potentially also be @safe and @nogc. I think that would resolve the linker errors.

They are all 3, see the top of the file. You need the pragmas, that's how LDC recognizes them as intrinsics.
